clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1455,8 +1455,604 @@ inline void segmented_sort_pairs_by_two_pair_sorts(
+          typename IdxHashFunc, typename... Range3>
+inline void __histogram_general_private_global_atomics(
+    Policy &&policy, Range1 &&input, Range2 &&bins, const Size &num_bins,
+    IdxHashFunc func, Range3 &&...opt_range) {
+
+  const ::std::size_t N = input.size();
+  auto __global_mem_size =
+      policy.queue()
+          .get_device()
+          .template get_info<sycl::info::device::global_mem_size>();
+  const ::std::size_t max_segments =
+      ::std::min(__global_mem_size / (num_bins * sizeof(BinType)),
+                 __ceiling_div(N, WorkGroupSize * __min_iters_per_work_item));
+  const ::std::size_t iters_per_work_item =
+      __ceiling_div(N, max_segments * WorkGroupSize);
+  ::std::size_t segments =
+      __ceiling_div(N, WorkGroupSize * iters_per_work_item);
+
+  sycl::buffer<BinType, 1> private_histograms(
+      sycl::range<1>(segments * num_bins));
+
+  auto e = policy.queue().submit([&](auto &h) {
+    // Temporary use of stable non-public API from oneDPL,  this function will
+    // be replaced with oneDPL call in an upcoming release.
+    oneapi::dpl::__ranges::__require_access(h, input, bins, opt_range...);
+    sycl::accessor hacc_private(private_histograms, h, sycl::read_write,
+                                sycl::no_init);
+    h.parallel_for(
+        sycl::nd_range<1>(segments * WorkGroupSize, WorkGroupSize),
+        [=](sycl::nd_item<1> __self_item) {
+          const ::std::size_t __self_lidx = __self_item.get_local_id(0);
+          const ::std::size_t __wgroup_idx = __self_item.get_group(0);
+          const ::std::size_t __seg_start =
+              WorkGroupSize * iters_per_work_item * __wgroup_idx;
+
+          __clear_wglocal_histograms(hacc_private, __wgroup_idx * num_bins,
+                                     num_bins, __self_item);
+          __accum_local_atomics<WorkGroupSize, ::std::uint32_t,
+                                sycl::access::address_space::global_space>(
+              input, __seg_start, N, __self_lidx, hacc_private,
+              __wgroup_idx * num_bins, func, __self_item, iters_per_work_item);
+
+          __reduce_out_histograms<BinType>(hacc_private,
+                                           __wgroup_idx * num_bins, bins,
+                                           num_bins, __self_item);
+        });
+  });
+  e.wait();
+}
+
+template <typename Policy, typename Iter1, typename Iter2, typename Size,
+          typename IdxHashFunc, typename... Range>
+inline Iter2
+__histogram_general_select_best(Policy &&policy, Iter1 first, Iter1 last,
+                                Iter2 histogram_first, const Size &num_bins,
+                                IdxHashFunc func, Range &&...opt_range) {
+  using __histo_value_type = typename ::std::iterator_traits<Iter2>::value_type;
+  auto __local_mem_size =
+      policy.queue()
+          .get_device()
+          .template get_info<sycl::info::device::local_mem_size>();
+  constexpr ::std::uint8_t __max_registers = 16;
```

Comment:
Do we know if there are any limitations with this hardcoded value for devices that may not have enough register space or should this be sufficient for most devices?

Reply:
It will result in slower code but not incorrect code, as it will spill to global memory.  It is currently 32 bytes, as each private bin is `std::uint16_t`.  

Interestingly, I saw decreased performance when using `std::uint8_t` register bins (tested in segment sizes which wouldn't possibly overflow).